lib/kitchen/driver/ec2.rb

```diff
@@ -194,84 +247,176 @@ def default_aws_session_token
+        require 'timeout'
+        @iam_creds ||= begin
+          timeout(5) do
+            Net::HTTP.get(URI.parse('http://169.254.169.254'))
+          end
+          fetch_credentials
+        rescue Errno::EHOSTUNREACH, Errno::EHOSTDOWN, Timeout::Error,
+          NoMethodError, ::StandardError => e
+          debug("fetch_credentials failed with exception #{e.message}:#{e.backtrace.join("\n")}")
+          {}
+        end
+      end
+
+      INSTANCE_METADATA_HOST = 'http://169.254.169.254'
+      INSTANCE_METADATA_PATH = '/latest/meta-data/iam/security-credentials/'
+      # fetch_credentials logic copied from Fog
+      def fetch_credentials
+        begin
+          connection = Excon.new(INSTANCE_METADATA_HOST)
+          role_name = connection.get(
+            :path => INSTANCE_METADATA_PATH, :expects => 200
+          ).body
+          role_data = connection.get(
+            :path => INSTANCE_METADATA_PATH+role_name, :expects => 200
+          ).body
+
+          session = MultiJson.load(role_data)
+          credentials = {}
+          credentials[:aws_access_key_id] = session['AccessKeyId']
+          credentials[:aws_secret_access_key] = session['SecretAccessKey']
+          credentials[:aws_session_token] = session['Token']
+          credentials[:aws_credentials_expire_at] = Time.xmlschema session['Expiration']
+          #these indicate the metadata service is unavailable or has no profile setup
+          credentials
+        rescue Excon::Errors::Error => e
+          info("Unable to fetch credentials: #{e.message}")
+          super
+        end
+      end
+
       private
 
-      def connection
-        Fog::Compute.new(
-          :provider               => :aws,
-          :aws_access_key_id      => config[:aws_access_key_id],
-          :aws_secret_access_key  => config[:aws_secret_access_key],
-          :aws_session_token      => config[:aws_session_token],
-          :region                 => config[:region],
-          :endpoint               => config[:endpoint],
+      # This copies transport config from the current config object into the
+      # state.  This relies on logic in the transport that merges the transport
+      # config with the current state object, so its a bad coupling.  But we
+      # can get rid of this when we get rid of these deprecated configs!
+      def copy_deprecated_configs(state)
+        if config[:ssh_timeout]
+          state[:connection_timeout] = config[:ssh_timeout]
+        end
+        if config[:ssh_retries]
+          state[:connection_retries] = config[:ssh_retries]
+        end
+        if config[:username]
+          state[:username] = config[:username]
+        elsif instance.transport[:username] == instance.transport.class.defaults[:username]
+          # If the transport has the default username, copy it from amis.json
+          # This duplicated old behavior but I hate amis.json
+          ami_username = amis['usernames'][instance.platform.name]
```

Comment:
We'll have to work this out longer term--ideally the Driver should be able to pick an appropriate default username (based on AMI assuming we have that info), but until then the behavior is maintained which is good.